iroh-net/src/relay/iroh_relay.rs

```diff
@@ -876,6 +876,164 @@ mod tests {
+            relay: Some(RelayConfig {
+                secret_key: SecretKey::generate(),
+                http_bind_addr: (Ipv4Addr::LOCALHOST, 0).into(),
+                tls: None,
+                limits: Default::default(),
+            }),
+            stun: None,
+            metrics_addr: None,
+        })
+        .await
+        .unwrap();
+        // NOTE: Using `ws://` URL scheme to trigger websockets.
+        let relay_url = format!("ws://{}", server.http_addr().unwrap());
+        let relay_url: RelayUrl = relay_url.parse().unwrap();
+
+        // set up client a
+        let a_secret_key = SecretKey::generate();
+        let a_key = a_secret_key.public();
+        let resolver = crate::dns::default_resolver().clone();
+        let (client_a, mut client_a_receiver) =
+            ClientBuilder::new(relay_url.clone()).build(a_secret_key, resolver);
+        let connect_client = client_a.clone();
+
+        // give the relay server some time to accept connections
+        if let Err(err) = tokio::time::timeout(Duration::from_secs(10), async move {
+            loop {
+                match connect_client.connect().await {
+                    Ok(_) => break,
+                    Err(err) => {
+                        warn!("client unable to connect to relay server: {err:#}");
+                        tokio::time::sleep(Duration::from_millis(100)).await;
+                    }
+                }
+            }
+        })
+        .await
+        {
+            panic!("error connecting to relay server: {err:#}");
+        }
+
+        // set up client b
+        let b_secret_key = SecretKey::generate();
+        let b_key = b_secret_key.public();
+        let resolver = crate::dns::default_resolver().clone();
+        let (client_b, mut client_b_receiver) =
+            ClientBuilder::new(relay_url.clone()).build(b_secret_key, resolver);
+        client_b.connect().await.unwrap();
+
+        // send message from a to b
+        let msg = Bytes::from("hello, b");
+        client_a.send(b_key, msg.clone()).await.unwrap();
+
+        let (res, _) = client_b_receiver.recv().await.unwrap().unwrap();
+        if let ReceivedMessage::ReceivedPacket { source, data } = res {
+            assert_eq!(a_key, source);
+            assert_eq!(msg, data);
+        } else {
+            panic!("client_b received unexpected message {res:?}");
+        }
+
+        // send message from b to a
+        let msg = Bytes::from("howdy, a");
+        client_b.send(a_key, msg.clone()).await.unwrap();
+
+        let (res, _) = client_a_receiver.recv().await.unwrap().unwrap();
+        if let ReceivedMessage::ReceivedPacket { source, data } = res {
+            assert_eq!(b_key, source);
+            assert_eq!(msg, data);
+        } else {
+            panic!("client_a received unexpected message {res:?}");
+        }
+    }
+
+    #[tokio::test]
+    async fn test_relay_clients_websocket_and_derp() {
```

Comment:
we should also have tests for the tls variants of these two tests

Reply:
This is not easy to do - mainly because `tokio_tungstenite_wasm::WebSocketStream` cannot be configured to trust self-signed certificates (in the browser you have no control over that!).

Eventually I want to create two code paths on the client side: One that uses `tokio_tungstenite_wasm` in the browser and one that uses `fasterwebsockets` or `tokio_tungstenite` directly with us having more control over it client-side. (Or what that basically means is vendoring `tokio_tungstenite_wasm` and making the stream behind it configurable & exposing a different way to establish connections more manually).

I could do this in this PR still, but it could totally be a separate PR. LMK what you prefer.
The PR would essentially enable (1) proxy support and (2) testing WSS locally and (3) turn more tests that currently only test the derp protocol into tests that test both.

Reply:
lets make it a follow up then

Reply:
Filed this: https://github.com/n0-computer/iroh/issues/2418